### ISSUE TITLE
SAM specification is overly strict on the value of FLG for BGZF headers

### ISSUE THREAD

**Issue by kortschak** [samtools/hts-specs#181]
The current version of the spec requires that FLG be set to 4. This is more strict that is required by RFC1952 which only requires that FLG.EXTRA must be set when the extra field contains data, which it does to hold the BGZF block size subfield. However, requiring that it be exactly 4 as the spec seems to prevents the use of FNAME and FCOMMENT which could otherwise be used.

See https://github.com/samtools/htslib/issues/451.

**Comment by jkbonfield** [comment]
Note that currently htslib implementation checks FLG&4, but this introduces bugs into the code.  Eg consider:

```echo -en '\x1f\x8b\x08\x0cdate\x02\x03\x06\x00BC\x02\x00\x1f\x00filename\x00\x4b\xcb\xcf\xe7\x02\x00\xa8\x65\x32\x7e\x04\x00\x00\x00' > _.bgz```

A straight bgzf_read() loop fails to decode this, but zcat does.  This is because the implementation has an inbuilt assumption of the gzip header being 18 bytes (ie no comment, no filename).  Technically this even means we could shoehorn in an entire new bgzf block and get different but legal decompression out of zcat vs our bgzf implementation if we managed to get a zlib block without nul chars.

While we could (and probably should) fix our implementation, I'm inclined to ask why not just fix it by forcing flags to be 4 instead, as per the spec itself.  I can see no benefit to allowing extra fields here and permitting implementations to be written easier with fixed sized headers seems like a good thing to promote.

**Comment by colinhercus** [comment]
I'll vote for this.

Colin

**Comment by kortschak** [comment]
The logical extension of that argument is that XLEN should be fixed at 0x6; it is currently free. One could fit a BGZF block in the extra field.

**Comment by jkbonfield** [comment]
I think that is fair too.  Many (all?) current implementations do not cope with anything other than XLEN as 6 anyway.   BGZF is a specialisation of Gzip so I don't see a problem with tightening it down further, especially when it is simply to match the limitations in existing implementations.

**Comment by kortschak** [comment]
Fair enough. Not all though - the Go implementation is completely concordant with the GZIP specification.

**Comment by innovate-invent** [comment]
I would ask that the BGZF standard be brought in line with the gzip standard to allow for extra functionality.
It would be nice if I could pack extra data in the block headers to transparently extend the format.
The implementations that currently expect a fixed size header are already incorrect by the fact that XLEN is free.

**Comment by jkbonfield** [comment]
May I ask what extra functionality you plan on adding?

We could update BGZF, but realistically it's hard to do given the huge number of out-dated implementations out there and a *massive* culture of copying-code with no desire to update.  (Do a github code search and you'll find hundreds and hundreds of copies of bits of samtools 0.1.19 copied  into various projects, including some mainstream ones.)

Therefore personally I'm really reluctant to loosen the spec unless there is a significant benefit to most users.  I'd be more inclined to tighten to and remove the "apparent" flexibility that in reality is problematic to use.

**Comment by innovate-invent** [comment]
I was toying with various ideas but the first that comes to mind is including a hint to the total decompressed size of all of the blocks in the document in the first block.

Another hint I would include would be a flag to guarantee that the current block contains only complete records. ie. A record is not split across multiple blocks.

It would also be nice to have a string with the id and version of the tool used to generate the file.
This way I can make some assumptions about the rest of the blocks in the file.

I had a few other things I wanted to do but generally it all comes down to being able to pack extra information into a compressed BAM that is accessible before decompressing that would not affect other (properly implemented)  tools.

**Comment by lh3** [comment]
I don't think we should touch the BGZF part. I agree the implementation should be more flexible if we started over, but now is far too late. Adding anything to the BGZF header means breaking all older samtools/htslib, which is hefty price we can't afford. This is not going to happen.
